substrate/client/transaction-pool/src/fork_aware_txpool/fork_aware_txpool.rs

```diff
@@ -0,0 +1,1591 @@
+							resubmitted_to_report += 1;
+
+							if !contains {
+								log::trace!(
+									target: LOG_TARGET,
+									"[{:?}]: Resubmitting from retracted block {:?}",
+									tx_hash,
+									hash,
+								);
+							}
+							!contains
+						})
+						.map(|(tx_hash, tx)| {
+							//find arc if tx is known
+							self.mempool.get_by_hash(tx_hash).unwrap_or_else(|| Arc::from(tx))
+						}),
+				);
+
+				self.metrics.report(|metrics| {
+					metrics.resubmitted_retracted_txs.inc_by(resubmitted_to_report)
+				});
+			}
+
+			let _ = view
+				.pool
+				.resubmit_at(
+					&hash_and_number,
+					// These transactions are coming from retracted blocks, we should
+					// simply consider them external.
+					TransactionSource::External,
+					resubmit_transactions,
+				)
+				.await;
+		}
+	}
+
+	/// Executes the maintainance for the finalized event.
+	///
+	/// Performs a house-keeping required for finalized event. This includes:
+	/// - executing the on finalized procedure for the view store,
+	/// - purging finalized transactions from the mempool and triggering mempool revalidation,
+	async fn handle_finalized(&self, finalized_hash: Block::Hash, tree_route: &[Block::Hash]) {
+		let finalized_number = self.api.block_id_to_number(&BlockId::Hash(finalized_hash));
+		log::debug!(target: LOG_TARGET, "handle_finalized {finalized_number:?} tree_route: {tree_route:?} views_count:{}", self.active_views_count());
+
+		let finalized_xts = self.view_store.handle_finalized(finalized_hash, tree_route).await;
+
+		self.mempool.purge_finalized_transactions(&finalized_xts).await;
+		self.import_notification_sink.clean_notified_items(&finalized_xts);
+
+		self.metrics
+			.report(|metrics| metrics.finalized_txs.inc_by(finalized_xts.len() as _));
+
+		if let Ok(Some(finalized_number)) = finalized_number {
+			self.revalidation_queue
+				.revalidate_mempool(
+					self.mempool.clone(),
+					HashAndNumber { hash: finalized_hash, number: finalized_number },
+				)
+				.await;
+		} else {
+			log::trace!(target: LOG_TARGET, "purge_transactions_later skipped, cannot find block number {finalized_number:?}");
+		}
+
+		self.ready_poll.lock().remove_cancelled();
+		log::trace!(target: LOG_TARGET, "handle_finalized after views_count:{:?}", self.active_views_count());
+	}
+
+	/// Computes a hash of the provided transaction
+	fn tx_hash(&self, xt: &TransactionFor<Self>) -> TxHash<Self> {
+		self.api.hash_and_length(xt).0
+	}
+}
+
+#[async_trait]
+impl<ChainApi, Block> MaintainedTransactionPool for ForkAwareTxPool<ChainApi, Block>
+where
+	Block: BlockT,
+	ChainApi: 'static + graph::ChainApi<Block = Block>,
+	<Block as BlockT>::Hash: Unpin,
+{
+	/// Executes the maintainance for the given chain event.
+	async fn maintain(&self, event: ChainEvent<Self::Block>) {
+		let start = Instant::now();
+		log::debug!(target: LOG_TARGET, "processing event: {event:?}");
+
+		self.view_store.finish_background_revalidations().await;
```

Comment:
Why are we finishing the background revalidations?

Reply:
To not block the maintain process.